Server side label selector feature

vinzent opened the pull request:
Allow to pass a labelSelector when creating a client.
    
The labelSelector is the string as used by the kubectl/oc `-l`
cli arg.
    
Example: create projects client
    
```
const projects_client = this.K8SClientFactory.create({
    kind: KubernetesAPI.WatchTypes.PROJECTS,
    labelSelector: "environment=dev"
});
```
    
Ref ENTESB-14539 and hawtio/hawtio-online#64

astefanutti commented:
I think we can assume no one is using the existing client-side filtering and re-use the `labelSelector` argument for server-side filtering.

vinzent commented:
Does it make sense to have the labelSelector feature on the get()/watch() methods? I think they return existing objects if a list already exists. If somebody passes a different selector it won't have an effect? 

Can i change the labelSecletor property to type string? leaving it LabelMap will required to decode the projectSelector string just to encode it later again.

astefanutti commented:
> Does it make sense to have the labelSelector feature on the get()/watch() methods? I think they return existing objects if a list already exists. If somebody passes a different selector it won't have an effect?

Right, it should be removed from the `get` and `watch` methods.

> Can i change the labelSelector property to type string? leaving it LabelMap will required to decode the projectSelector string just to encode it later again.

Yes, let's change it to `string`. We can migrate to a structured declaration later on.

vinzent committed:
Server side filtering by label selector

Allow to pass a labelSelector when creating a client.

The labelSelector is the string as used by the kubectl/oc `-l`
cli arg.

Example: create projects client

```
const projects_client = this.K8SClientFactory.create({
  kind: KubernetesAPI.WatchTypes.PROJECTS,
  labelSelector: "environment=dev"
});
```

Ref ENTESB-14539 and hawtio/hawtio-online#64

vinzent commented on plugins/kubernetes-api/ts/client.ts:
```diff
@@ -546,39 +546,56 @@ module KubernetesAPI {
-        if (location && url.indexOf("://") < 0) {
-          var hostname = location.hostname;
+        let urlStr = UrlHelpers.join(masterApiUrl(), this._path);
+        let location = Core.windowLocation();
```
wondering why this construct is used here but not in the `_restUrl` method.

astefanutti replied:
I'm inclined to think that it is not used. My guess would be that the WebSocket API may not work well with relative path, while the polling client does. Anyway, we can leave it for further analysis.

astefanutti commented on plugins/kubernetes-api/ts/kubernetesApiHelpers.ts:
```diff
@@ -268,16 +268,6 @@ module KubernetesAPI {
-  /**
-   * Filter a collection by label
-   **/
-  export function filterByLabel(objects: Array<any>, labelSelector: LabelMap) {
```
The `labelsToString` and `selectorMatches` methods below may also be removed.

astefanutti replied:
The `labelFilterTextSeparator` variable may also be removed.

vinzent replied:
removed

vinzent committed:
Remove unused helpers and vars

Due to removal of the post label filtering, these methods and
vars are not required anymore.

astefanutti commented:
Thanks! Let me cut a new release.
